brave/src/test/java/brave/TracerTest.java

```diff
@@ -693,6 +695,52 @@
         .isPositive();
   }
 
+  @Test public void useSpanAfterFinished_doesNotCauseBraveFlush() throws InterruptedException {
+    simulateInProcessPropagation(tracer, tracer);
+    blockOnGC();
+    tracer.newTrace().start().abandon(); //trigger orphaned span check
+    assertThat(spans).hasSize(1);
+    assertThat(spans.stream()
+      .flatMap(span -> span.annotations().stream())
+      .map(Annotation::value)
+      .collect(Collectors.toList())).doesNotContain("brave.flush");
+  }
+
+  @Test public void useSpanAfterFinishedInOtherTracer_doesNotCauseBraveFlush()
+    throws InterruptedException {
+    Tracer noOpTracer = Tracing.newBuilder()
+      .build().tracer();
+    simulateInProcessPropagation(noOpTracer, tracer);
+    blockOnGC();
+    tracer.newTrace().start().abandon(); //trigger orphaned span check
+
+    // We expect the span to be reported to the NOOP reporter, and nothing to be reported to "spans"
+    assertThat(spans).hasSize(0);
+    assertThat(spans.stream()
+      .flatMap(span -> span.annotations().stream())
+      .map(Annotation::value)
+      .collect(Collectors.toList())).doesNotContain("brave.flush");
+  }
+
+  /**
+   * Must be a separate method from the test method to allow for local variables to be garbage
+   * collected
+   */
+  private static void simulateInProcessPropagation(Tracer tracer1, Tracer tracer2) {
```

Comment:
this is an odd use case. the tracer (tracing etc) is a singleton and meant to be used as such.

In-process propagation involves moving the Span or TraceContext across threads, not switching tracer instances per thread.

Can you elaborate why you are doing this? It might cause other problems then the scenario noted here.

Reply:
I was not aware that the tracer/tracing was meant to be a singleton. The API lets one create different instances via Tracing.newBuilder(). Is the intent that this builder is used only once per application? If so, why does the API allow you to create multiple instances?

As a little bit of background, we have different tracing instances each with a different serviceName. As an example lets say I have webservice A<sub>ws</sub> and webservice B<sub>ws</sub>. A<sub>ws</sub> requests information from B<sub>ws</sub> through a library B<sub>client</sub> that makes http calls to B<sub>ws</sub>. The B<sub>client</sub> library uses a tracing instance with serviceName "B" whereas the A<sub>ws</sub> has a tracing instance with serviceName "A". A trace will start in A<sub>ws</sub> with a span that has service name "A". This service will eventually call into the B<sub>client</sub> library code which will create child spans (via a different tracer) with serviceName "B".

We've actually had some discussions internally recently on whether or not we should be doing this, so some guidance on when different tracing instances should be used would be appreciated. 

As for the in-process propagation, I think you might be confused that the test does not actually use multiple threads. I found that this is not strictly necessary because the behavior is actually caused by using multiple tracing instances and scoped spans. I just found the in-process propagation example to be an easy one to understand.

Reply:
I will try to answer your questions:

Q: why let people make instances if should be a singleton?
A: many frameworks handle singletons differently, and there are classloader concerns, too. brave does not constrain you to using spring, guice, or otherwise singleton management.

Q: should we be using different instances per endpoint in order to control the service name?
A: this is complex and likely more fine grained than most use. For most sites endpoints are actually span names. We have a tool called FinishedSpanHandler which has the concept of localRoot which would allow you to rewrite data consistently per fraction of a trace. Again, this is more complex that what most sites do. if you want to discuss that further better use chat as it is less spammy to watchers https://gitter.im/openzipkin/zipkin

On the test talking about threads then not using them etc. I personally found this confusing and it took me a few hours to guess what was supposed to be checked. (think time not pure work). In general, being precise like your diagram is best as when comments lie they are distracting to current and future maintainers. Your tests helped, just it mentioned things that wouldn't have been tested in that code. thx regardless!

Reply:
Thanks for the answers! This is super helpful. And sorry for any confusion those tests caused.